### PR TITLE
Use new Pages API over imperative navigator api

### PR DESCRIPTION
**Pull request by Renzo-Olivares**
This change replaces `namedRoutes` style of navigation which is tied to the old navigator api, with the new `Router` & `Pages` method of navigation. This gives us tons of more flexibility with our navigator stack, and allows us to navigate through screen declaratively, by reacting to changes in our apps state. This is different than the old way of directly popping and pushing routes onto the navigator stack. Here we are directly rebuilding our navigator with the pages and transitions that we want based on our applications state.

Issues fixed:
* SharedAxisTransition now works correctly with the new Pages API. SearchPage is now its own `Page` and we navigate to it using our navigator, rather than using `PageTransitionSwitcher` to swap between the `HomePage` and the `SearchPage`
* Navigating through Mailbox destinations is now reactive, so there are no delays in the fade through animation. Before we would delay our `setState` call to wait for our drawer to animate completely and then we would rebuild our `MailNavigator` with a new `InboxPage` to enable the fade through transition. Now our `MailViewRouterDelegate` listens for changes in our state and rebuilds our navigator with the correct `InboxPage` destination at the top of the navigator page stack, with the fade through transition.
* Calling the back button while the bottom drawer is visible will dismiss the drawer instead of closing the whole app.
* Back to home navigation : Calling the back button on a mailbox that is not the inbox will send you back to the inbox instead of closing the app.

Changes Implemented:
* `RouterProvider` to keep track of current route in the root router
* `ReplyRouterDelegate` that listens to `RouterProvider` for changes in `routePath`, and returns a `Navigator` with the desired configuration. Handles `HomePage` and `SearchPage` routes.
* `SharedAxisTransitionPageWrapper` that returns a `Page` that transitions with the `SharedAxisTransition`
* `FadeThroughTransitionWrapper` that returns a `Page` that transition with the `FadeThroughTransition`
* `ReplyRoutePath` and its children classes to define routes for the root router
* A simple `ReplyRouteInformationParser` to parse the route information and send to the `RouterDelegate` for configuration.
* Simple `MailViewRouterDelegate` that listens to `EmailStore` for changes in the `currentlySelectedInbox`, and returns a `Navigator` with the correct configuration, in this case an `InboxPage` wrapped in a `FadeThroughTransitionPageWrapper`. Handles `InboxPage` route. This one is simpler compared to `ReplyRouterDelegate` because it only handles one `Page`. While it does handle more than one route, those routes are ones produced by the `OpenContainer` widget, so they are not accessible except through the old imperative navigator api. This is where our `popRoute` logic comes in.
* `popRoute` in `MailViewRouterDelegate` handles most of the back button calls since most of the apps flow is within the inside navigator. This is where all the back button calls of the inner navigator go, and we decide how to handle them, whether that is through changing state, or through the old navigator api. 

More on new [`Router and Pages API`](https://docs.google.com/document/d/1Q0jx0l4-xymph9O6zLaOY4d_f7YFpNWX_eGbzYxr9wY/edit#heading=h.eqh5rb9rdbai)

Fade through delay and back to home| Bottom drawer dismiss on back
---|---
reply-fade-thru-delay-fix|reply-drawer-dimisss

### CODE REVIEW

File: lib/home.dart [hunk 1]
```diff
@@ -600,19 +563,16 @@ class _BottomDrawerDestinations extends StatelessWidget {
...
         assert(dropArrowController != null),
-        assert(selectedIndex != null),
         assert(onItemTapped != null);
 
   final List<_Destination> destinations;
   final AnimationController drawerController;
   final AnimationController dropArrowController;
-  final int selectedIndex;
-  final void Function(int, String) onItemTapped;
+  final void Function(String) onItemTapped;
```
Comment (JoseAlba) [inline lib/home.dart:575]: You can use a typedef for this or just use ValueChanged instead. A lot of other classes use the ValueChanged parameter instead and I think we should try to keep one parameter consistent 

typedef ValueChanged<T> = void Function(T value);

File: lib/mail_view_router.dart [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+      }
+
+      if (emailStore.currentlySelectedInbox != 'Inbox') {
+        emailStore.currentlySelectedInbox = 'Inbox';
+        return SynchronousFuture<bool>(true);
+      }
+      return SynchronousFuture<bool>(false);
+    }
+
+    //Handles the back button when on the ComposePage.
```
Comment (JoseAlba) [inline lib/mail_view_router.dart:68]: space after comment

Comment (JoseAlba) [reply lib/mail_view_router.dart:68]: ComposePage => [ComposePage]

File: lib/mail_view_router.dart [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+          .currentlySelectedEmailId = -1;
+      return SynchronousFuture<bool>(true);
+    }
+
+    return SynchronousFuture<bool>(false);
+  }
+
+  @override
+  Future<void> setNewRoutePath(void configuration) {
+    // should never be called
```
Comment (JoseAlba) [inline lib/mail_view_router.dart:97]: Full sentence and this method doesn't seem necessary unless it is a required override.

File: lib/mail_view_router.dart [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    throw UnimplementedError();
+  }
+}
+
+class FadeThroughTransitionPageWrapper extends TransitionBuilderPage {
+  FadeThroughTransitionPageWrapper(
+      {@required this.child, @required this.transitionKey})
```
Comment (JoseAlba) [inline lib/mail_view_router.dart:104]: add comma for last parameter and run flutter format .

File: lib/router.dart [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+  @override
+  void dispose() {
+    replyState.removeListener(notifyListeners);
+    super.dispose();
+  }
+
+  @override
+  ReplyRoutePath get currentConfiguration =>
+      replyState.routePath; //appState.routePath
```
Comment (JoseAlba) [inline lib/router.dart:36]: Comment should be above the override and a complete sentence.

File: lib/router.dart [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+  }
+
+  bool _handlePopPage(Route<dynamic> route, dynamic result) {
+    // _handlePopPage should not be called on the home page because the
+    // PopNavigatorRouterDelegateMixin will bubble up the pop to the
+    // SystemNavigator if there is only one route in the navigator.
+    assert(route.willHandlePopInternally ||
+        replyState.routePath is ReplySearchPath);
+
+    final bool success = route.didPop(result);
```
Comment (JoseAlba) [inline lib/router.dart:74]: success might not be the best variable name

**Review by JoseAlba — COMMENTED** [review]
Looks really good

File: lib/mail_view_router.dart [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+    //Handles the back button when on the ComposePage.
+    if (onCompose) {
+      return SynchronousFuture<bool>(false);
+    }
+
+    // Handles the back button when the bottom drawer is visible on the
+    // MailView. Dismisses the drawer on back button press.
+    if (emailStore.bottomDrawerVisible) {
+      if (onMailView) {
```
Comment (rami-a) [inline lib/mail_view_router.dart:76]: Reduce nesting:
```suggestion
    if (emailStore.bottomDrawerVisible && onMailView) {
```

File: lib/mail_view_router.dart [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+              fillColor: Theme.of(context).scaffoldBackgroundColor,
+              animation: animation,
+              secondaryAnimation: secondaryAnimation,
+              child: child,
+            );
+          },
+        );
+
+  final Widget child;
+  final ValueKey transitionKey;
```
Comment (rami-a) [inline lib/mail_view_router.dart:120]: If these are just being passed to the super constructor, do you even need these properties?

File: lib/router.dart [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+const String _searchPageLocation = '/reply/search';
+
+class ReplyRouterDelegate extends RouterDelegate<ReplyRoutePath>
+    with ChangeNotifier, PopNavigatorRouterDelegateMixin<ReplyRoutePath> {
+  @override
+  final GlobalKey<NavigatorState> navigatorKey;
+
+  RouterProvider replyState;
+
+  ReplyRouterDelegate({@required this.replyState})
```
Comment (rami-a) [inline lib/router.dart:20]: Put the constructor before the properties to be consistent

File: lib/router.dart (hunk 5 above)
Comment (rami-a) [reply lib/router.dart:36]: This comment doesn't seem to add additional info, maybe get rid of it?

File: lib/router.dart [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+              animation: animation,
+              secondaryAnimation: secondaryAnimation,
+              transitionType: SharedAxisTransitionType.scaled,
+              child: child,
+            );
+          },
+        );
+
+  final Widget child;
+  final ValueKey transitionKey;
```
Comment (rami-a) [inline lib/router.dart:119]: Same comment as above, do you need these?

**Review by rami-a — APPROVED** [review]

File: lib/mail_view_router.dart (hunk 3 above)
Comment (Renzo-Olivares) [reply lib/mail_view_router.dart:97]: It is a required override.

File: lib/router.dart (hunk 5 above)
Comment (Renzo-Olivares) [reply lib/router.dart:36]: Yeah your right I forgot to get rid of this.

File: lib/router.dart (hunk 6 above)
Comment (Renzo-Olivares) [reply lib/router.dart:74]: changed to `didPop`

File: lib/home.dart (hunk 1 above)
Comment (Renzo-Olivares) [reply lib/home.dart:575]: Done!

File: lib/mail_view_router.dart (hunk 8 above)
Comment (Renzo-Olivares) [reply lib/mail_view_router.dart:119]: mm I don't think I know what you mean. I am still passing these properties to the super constructor, so I think I still need them.

File: lib/router.dart (hunk 10 above)
Comment (Renzo-Olivares) [reply lib/router.dart:118]: replied above

**Review by JoseAlba — APPROVED** [review]
```
██╗      ██████╗ ████████╗███╗   ███╗
██║     ██╔════╝ ╚══██╔══╝████╗ ████║
██║     ██║  ███╗   ██║   ██╔████╔██║
██║     ██║   ██║   ██║   ██║╚██╔╝██║
███████╗╚██████╔╝   ██║   ██║ ╚═╝ ██║
╚══════╝ ╚═════╝    ╚═╝   ╚═╝     ╚═╝
```

**Pull request merged**